### DISCUSSION TITLE
[Java client querying] Order by count() without having to select count()

### DISCUSSION THREAD

**Discussion by Honeyxilia** [ravendb/ravendb#18601] [Q&A]
hello !

I'm trying to do a query with a group by/order by count combo - for starters, here are the relevant tables : 

```java
class UserRaven {
    private String id;
	private String username;
	private LocalDate creationDate;
	private String gender;
}

class BorrowRaven {
    	private String id;
	private MovieRaven movie;
	private UserRaven user;
	private LocalDate borrowDate;
}
```

the query i'm struggling with is finding the User that has the most Borrows - in RQL, it's the following query : `from "BorrowRavens" group by user order by count() desc select user`

Here's the result on Studio : 
![image](https://github.com/ravendb/ravendb/assets/72192278/5d0bd3be-5be7-4ada-8440-fcb37ebb893b)

However, I'm trying to integrate it in Java with the following query : 
```java
User user = session.query(BorrowRave,.class)
    .groupBy("user')
    .selectKey("user()
    .selectCount()
    .orderByDescending("count")
    .ofType(User.Ravenclass)
    .first()
```

However, that statement corresponds to the following RQL command : 
`from "BorrowRavens" group by user order by count desc select count() as count, user`

My issue here is that i wish to only obtain the User class, yet the `selectCount()` does order by count of users, but also puts the count value in the select fields, making the projection not work correctly...

Note: i tried doing a rawQuery : `session.advanced().rawQuery(UserRaven.class, "from 'BorrowRavens' group by user order by count() select user").first()`, but while the query works on the website, it only returns a User with null fields...

Is there a solution querying that user without the count value ? Alternatively, is it possible to remove count as a selected field afterwards ? ty!!

**Comment by ayende** [comment]
That is a really strange way to query. Your `user` field is an object? 
Shouldn't you be grouping by `user.id`, instead?

The problem in your query is that you are using ` .orderByDescending("count")`, and that sends `count` to the server.
You need `count()`, instead. The issue is that you cannot pass it via this method, since it would escape it.

Also note that you need to order using numeric and not string comparison.

Can you use this query:

```
from "BorrowRavens" 
group by user.id
order by count() as long desc 
select user, count()
include user
```

Then you have everything in one go, and the extra `count()` field doesn't matter

**Reply by Honeyxilia** [reply]
hi, thanks for the response and for your time !

i tried to do that request, and unfortunally, grouping directly by the `user.id` only return a `user.id` column, which would then require another query or loading to get to the actual user... and grouping my user allows me to get the whole user straight away c:

about the `count` variable, it is not added by the `orderByDescending()` method, but by the `selectCount` method - for example, the following line : 
```java
session.query(BorrowRaven.class)
    .groupBy("user")
    .selectKey("user")
    .selectCount(); // used for aggregation by count
```

when accessed with a debugger shows that it correponds to the following RQL query : 
```
from 'BorrowRavens' group by user select user, count() as count 
```

So it seems that `selectCount` is the method that creates the count value which is then used in the `orderBy` method - however my question is to remove the count value in select :/

is it possible to remove the newly created `count` field by `selectCount()` from the select part?
ty!

**Reply by ayende** [reply]
Note that with this query:

```
from "BorrowRavens" 
group by user.id
order by count() as long desc 
select user, count()
include user
```

You are going to the server _once_, fetching both the user ids _and_ the associated users.

See:
https://ravendb.net/docs/article-page/6.0/java/client-api/session/loading-entities#load-with-includes

And about the API, you are correct, `selectCount()` add it, but without calling it, you cannot `order by count()` through the API.

What I think you want is something like this:

```
IdAndCountDto dto = session.query(IdAndCountDto,.class)
    .groupBy("user.id")
    .selectKey("user.id")
    .selectCount()
    .orderByDescending("count")
    .include("user.id")
    .first()
User user = session.load(User.clas, dto.id);
```

That has a _single_ query to the database, and gets the data
